### PR TITLE
add support to metrics calculation

### PR DESCRIPTION
**Pull request by TaekyungHeo**
# summary
Add support to metrics calculation.
1. Iteration E2E time;
2. bandwidth;

# test plan
```
srun -l -N 8 -n 64 --ntasks-per-node=8 --container-name=sanshang_test \
--container-mounts=/labhome/sanshang:/labhome/sanshang,/swgwork/sanshang:/labhome/sanshang/storage \
bash -c "comm_replay --trace-type et --trace-path ${ROOT_PATH}/et_trace_64gpus \
--output-path ${ROOT_PATH}/output \
--enable-profiler --profiler-num-replays-start 4 --profiler-num-replays 5 \
--do-warm-up --reuse-tensors --num-replays 10"
```
image
image

### CODE REVIEW

File: et_replay/comm/profiler_trace_analysis.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    
+    group_size = _get_dict_value(evt['args'], 'Group size', f'Missing "Group size" in event: {evt}')
+    correction_factor_func = _get_dict_value(_collname_to_busbw_corr_factor_func,
+                                             coll_name, 
+                                             f'Unsupported collective op for busbw calculation: {coll_name}')
+    
+    return correction_factor_func(group_size)
+
+def calculate_bw_(trace_data):
+    nccl_events = [i for i in trace_data['traceEvents'] if i.get('cat', '') == 'kernel' and i['name'].startswith('ncclDevKernel_')]
```
Comment (briancoutinho) [inline et_replay/comm/profiler_trace_analysis.py:113]: nit could be startswith('nccl')  for older nccl kernel names that did not start with Device

Comment (briancoutinho) [inline et_replay/comm/profiler_trace_analysis.py:113]: to save memory consider using generator
nccl_events = (e for e in trace_data['traceEvents'] if e.get('cat', '') == 'kernel' and e['name'].startswith('ncclDevKernel_'))

File: et_replay/comm/profiler_trace_analysis.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+            
+            algbw = _calculate_algbw(evt)
+            busbw_factor = _get_event_busbw_factor(evt)
+            busbw = round(algbw * busbw_factor, 2)
+
+            evt['args']['algbw (GB/sec)'] = algbw
+            evt['args']['busbw (GB/sec)'] = busbw
+            evt['args']['busbw_factor'] = busbw_factor
+        except ValueError as e:
+            logger.error('Error processing event: %s', e)
```
Comment (briancoutinho) [inline et_replay/comm/profiler_trace_analysis.py:130]: we could events processing that failed and print at bottom

File: et_replay/comm/profiler_trace_analysis.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+def calculate_sbw(trace_data):
+    # calculate shared bw per rank
+    nccl_events = [i for i in trace_data['traceEvents'] if i.get('cat', '') == 'kernel' \
+                   and i['name'].startswith('ncclDevKernel_') \
+                   and 'busbw_factor' in i['args']]
+    
+    if not len(nccl_events):
+        return 0
+    
+    total_data_size = sum([_calculate_event_data_size(evt) * _get_event_busbw_factor(evt) for evt in nccl_events])
```
Comment (briancoutinho) [inline et_replay/comm/profiler_trace_analysis.py:141]: nit this will work
total_data_size = sum(_calculate_event_data_size(evt) * _get_event_busbw_factor(evt) for evt in nccl_events)

File: et_replay/comm/profiler_trace_analysis.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+    end_time_point = max([i.end for i in time_range_tree])
+
+    sorted_tr = sorted(time_range_tree)
+    total_idle_time = \
+        sum([sorted_tr[i + 1].begin - sorted_tr[i].end for i in range(len(sorted_tr) - 1)]) \
+        if len(sorted_tr) > 1 else 0
+
+    return total_data_size / (end_time_point - begin_time_point - total_idle_time) / 1e3
+
+def pick_iter_e2e_time_(trace_data, tl):
```
Comment (briancoutinho) [inline et_replay/comm/profiler_trace_analysis.py:156]: please add some doc on this function, not clear what it is doing, and what is tl

File: et_replay/comm/profiler_trace_analysis.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+    nccl_events = [i for i in trace_data['traceEvents'] if i.get('cat', '') == 'kernel' \
+                   and i['name'].startswith('ncclDevKernel_') \
+                   and 'busbw_factor' in i['args']]
+    
+    if not len(nccl_events):
+        return 0
+    
+    total_data_size = sum([_calculate_event_data_size(evt) * _get_event_busbw_factor(evt) for evt in nccl_events])
+
+    time_range_tree = IntervalTree([Interval(evt['ts'], evt['ts'] + evt['dur']) for evt in nccl_events])
```
Comment (briancoutinho) [inline et_replay/comm/profiler_trace_analysis.py:143]: Today i learned Interval tree class nice :)

File: et_replay/comm/profiler_trace_analysis.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+                              np.average(busbw_lst),
+                              np.percentile(busbw_lst,  1),
+                              np.percentile(busbw_lst, 50),
+                              np.percentile(busbw_lst, 90),
+                              np.percentile(busbw_lst, 99)]
+    comm_bw_summary = dict(sorted(comm_bw_summary.items()))
+
+    # dump summary report
+    with open(os.path.join(report_dir, 'profiler_trace_summary_report.txt'), 'w', encoding='utf-8') as f:
+        f.write(f'avg. E2ETime of iters among all ranks: {sum(iter_e2e_time) / len(iter_e2e_time) / 1e3 :.3f} ms\n')
```
Comment (briancoutinho) [inline et_replay/comm/profiler_trace_analysis.py:232]: You can consider using tabulate library since you are considering adding dependency.

**Review by briancoutinho — COMMENTED** [review]
Left some suggestions. @shengfukevin  also saw black formatting issues, which we would need to fix to pass linter, Sheng is there someway we can share the lint suggestions? Let's do that after next iteration of the PR

File: et_replay/comm/backend/pytorch_dist_backend.py [hunk 7]
```diff
@@ -622,6 +622,9 @@ def barrier(self, collectiveArgs, name="dummy", retFlag=False):
 
         if retFlag:
             return retObj
+        
+    def barrier_all_ranks(self):
+        dist.barrier(device_ids=[self.get_device().index] if dist.get_backend() == "nccl" else None)
```
[the remark this answers is not in the document]
Comment (sanshang-nv) [reply et_replay/comm/backend/pytorch_dist_backend.py:627]: This is new added util function, not exist before. To make all dist API call be encapsulated in backend module.

**Comment by shengfukevin** [comment]
@briancoutinho @sanshang-nv @TaekyungHeo , I copied this PR and imported to Meta, this is the new diff https://github.com/facebookresearch/param/pull/196. Please update on PR 196. 

I have two issues that need fix:
1. When profiler is on, this PR will create two profilers, one for meta internal, one for OSS. I fixed that in PR196. 
2. Meta internal need to support uploading report to a Meta internal repository when --output is in the options. I am still working on it.

File: et_replay/comm/profiler_trace_analysis.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+    'Float8_e4m3fn'   : 1,
+    'Float8_e5m2fnuz' : 1,
+    'Float8_e4m3fnuz' : 1,
+    'UInt16'          : 2,
+    'UInt32'          : 4,
+    'UInt64'          : 8,
+}
+
+# refer to: https://github.com/NVIDIA/nccl-tests/blob/master/doc/PERFORMANCE.md
+_busbw_correction_factors_tbl: Dict[str, Callable[[int], float]] = {
```
[the remark this answers is not in the document]
Comment (sanshang-nv) [reply et_replay/comm/profiler_trace_analysis.py:50]: what's "nit"? Not get your point. Could please elaborate more? @briancoutinho

File: et_replay/comm/profiler_trace_analysis.py (hunk 1 above)
Comment (sanshang-nv) [reply et_replay/comm/profiler_trace_analysis.py:125]: You mean, older nccl kernel in latest nccl verion? Is there some examples? I don't assume we need to be compatible with old nccl version. But if new nccl version has older nccl kernels, we need to consider refine this.

**Comment by sanshang-nv** [comment]
>> When profiler is on, this PR will create two profilers, one for meta internal, one for OSS. I fixed that in PR196.
@shengfukevin what do you mean 2 profilers? When meta internal works, I will skip to not make public profiler work.

**Comment by sanshang-nv** [comment]
PR196 seems not include this commit: 57f619e0d84bda5f6917bbaba85ab8b7d5fbc541

File: et_replay/comm/profiler_trace_analysis.py (hunk 2 above)
Comment (sanshang-nv) [reply et_replay/comm/profiler_trace_analysis.py:142]: Done

File: et_replay/comm/profiler_trace_analysis.py (hunk 3 above)
Comment (sanshang-nv) [reply et_replay/comm/profiler_trace_analysis.py:153]: Done

File: et_replay/comm/profiler_trace_analysis.py (hunk 4 above)
Comment (sanshang-nv) [reply et_replay/comm/profiler_trace_analysis.py:168]: done

**Comment by sanshang-nv** [comment]
Hi, @shengfukevin 
Has updated to:
1. solve your 1st issue of 2 profilers when running inside Meta
2. resolve some comments from Brian.

**Comment by shengfukevin** [comment]
@sanshang-nv, after took another look, MetaProfiler seems an overkill to provide a Meta internal profiler. The difference between MetaProfiler and public profiler are 1. the supported activities 2. the call back function when trace is ready. Inside Meta, we already have fb directory, it seems redundant to have vendor_internal directory.  

When you create the profiler, can you just pass in different activities and callback function like the following:

try:
    from param_bench.et_replay.fb.internals import (
        get_fb_profiler_activities,
        get_fb_profiler_trace_handler,
    )

    has_internal_libs = True
except ImportError:
    has_internal_libs = False
    
if has_internal_libs:
        activities = get_fb_profiler_activities(device)
        trace_handler = get_fb_profiler_trace_handler(rank)
    else:
        activities = {ProfilerActivity.CPU, ProfilerActivity.CUDA}

        def trace_handler(p):
            import pathlib

            folder_path = os.path.join(output_path, "profiler_trace")
            try:
                pathlib.Path(folder_path).mkdir(parents=True, exist_ok=True)
            except PermissionError:
                logger.error(f"Permission denied to create directory {folder_path}")
            p.export_chrome_trace(os.path.join(folder_path, f"rank-{rank}.json"))

    logger.debug("GPU Trace Collection: Enabled")
    _torch_profiler = torch.profiler.profile(
        schedule=torch.profiler.schedule(
            wait=1,
            warmup=numWarmupIters,
            active=numIters,
            repeat=1,
        ),
        on_trace_ready=trace_handler,
        activities=activities,
    )
    
    Also I think we still need warmup-iter so user can control how many iterations they want to run the warmup.

Thanks
Sheng

**Comment by sanshang-nv** [comment]
> Inside Meta, we already have fb directory, it seems redundant to have vendor_internal directory.

I don't think so. `vendor_internal` is not only for Meta, but also for NV and other companies, to decouple vendor's internal features.

> param_bench.et_replay.fb.internals

I don't find `fb.internals` module in `param\et_replay` directory. Is it Meta private? I can just use it as below:
```
from param_bench.et_replay.fb.internals import (
get_fb_profiler_activities,
get_fb_profiler_trace_handler,
)
```

Then, we remove the usage to these 3 fb functions, right?
```
    from param_bench.train.comms.pt.fb.internals import (
        fbInitProfiler,
        fbSampleProfiler,
        fbStartProfiler,
    )
```

Because this is black box to me, it's hard to decide which implementation is the best from my side.

**Comment by shengfukevin** [comment]
ok, then let's keep vendor_internal. fb.internal got filtered out when export param from meta. 

You can call

from param_bench.et_replay.fb.internals import (
get_fb_profiler_activities,
get_fb_profiler_trace_handler,
)

inside vendor_internal/fb_internal.py 

Remove fbInitProfiler,
        fbSampleProfiler,
        fbStartProfiler,
        
Thanks
